server/controllers/shiftController.js

```diff
@@ -122,82 +122,93 @@ let ShiftController = function (
+  this.cancel = async function (req, res) {
     const errors = validationResult(req);
     if (!errors.isEmpty()) {
       return res.status(400).json({message: "Invalid request", errors: errors.array()});
     }
 
+    // Check there is a booking
+    let booking;
+    try {
+      booking = await bookingRepository.getById(req.params.id, req.user.id);
+    } catch (err) {
+      return res.status(500).json({message: errorMessage(err)});
+    }
+    if (!booking) {
+      return res.status(400).json({message: "No such booking exists"});
+    }
+
+    let shift = booking.shift;
+
+    // Check the user is a volunteer
+    let volunteer;
+    try {
+      volunteer = await volunteerRepository.getById(req.user.id);
+    } catch (err) {
+      return res.status(500).json({message: errorMessage(err)});
+    }
+    if (!volunteer) {
+      return res.status(400).json({message: "Only a volunteer can cancel a booking"});
+    }
+
+    // Check there is a shift creator
     let creator;
-    let shift;
+    try {
+      creator = await adminRepository.getById(shift.creatorId);
+    } catch (err) {
+      return res.status(500).json({message: errorMessage(err)});
+    }
+    if (!creator) {
+      return res.status(400).json({message: "Shift has no creator"});
+    }
 
-    bookingRepository
-      .getById(req.params.id, req.user.id)
-      .then(booking => {
-        if (!booking) {
-          res.status(400).json({message: "No such booking exists"});
-        } else {
-          shift = booking.shift;
-          return adminRepository.getById(shift.creatorId);
-        }
-      })
-      .then(admin => {
-        if (!admin) {
-          res.status(400).json({message: "Shift has no creator"});
-        } else {
-          creator = admin;
-          return volunteerRepository.getById(req.user.id);
-        }
-      })
-      .then(volunteer => {
-        if (!volunteer) {
-          res
-            .status(400)
-            .json({message: "Only a volunteer can cancel a booking"});
-        } else {
-          const message = constructCancelBookingMessage(
-            creator,
-            volunteer,
-            shift,
-            req.body.reason
-          );
-          if (creator.user.contactPreferences.email) {
-            let emailClient = new EmailClient(emailClientTypes.NOREPLY);
-            emailClient.send(creator.user.email, "Cancelled booking", message);
-          }
-          if (creator.user.contactPreferences.text) {
-            let smsClient = new SMSClient();
-            smsClient.send(creator.user.telephone, message);
-          }
-          return bookingRepository.delete(req.params.id, req.user.id);
-        }
-      })
+    // Send message to creator and cancel shift
+    const message = constructCancelBookingMessage(
+      creator,
+      volunteer,
+      shift,
+      req.body.reason
+    );
+    if (creator.user.contactPreferences.email) {
+      let emailClient = new EmailClient(emailClientTypes.NOREPLY);
```

Comment:
nit: `emailClient` and `smsClient` should be `const`.

Reply:
Fixed in next commit